File: internal/graphql/lint/naming_convention.go
// Copyright 2026 Outreach Corporation. All Rights Reserved.

// Description: Tier 3 naming-convention rule.

// naming_convention.go implements 1 (of 10) Tier 3 custom rules:
// naming-convention, which requires names to follow the casing,
// prefix/suffix, and underscore conventions its options configure per
// kind of definition, matching @graphql-eslint's rule of the same
// name.
//
// @graphql-eslint's own naming-convention accepts an option key for
// any AST node kind, including operation-only kinds (OperationDefinition,
// FragmentDefinition, VariableDefinition) that never appear in a
// standalone schema file, and lets a key carry an arbitrary selector
// (for example "FieldDefinition[parent.name.value!=Query]" -- see
// https://eslint.org/docs/developer-guide/selectors). devbase graphql
// lint only ever parses schema files (see lint.go's Files), so this
// port narrows the recognized keys to the schema-only kinds: the 6
// type-definition kinds (via the shared "types" option and its
// per-kind overrides), FieldDefinition, InputValueDefinition (covers a
// field's or directive definition's own argument, and an input object
// field -- graphql-js represents all three as the one AST kind),
// EnumValueDefinition, and DirectiveDefinition. It also recognizes the
// 3 root-operation-type field selectors from @graphql-eslint's own
// documented "recommended schema" config
// (FieldDefinition[parent.name.value=Query|Mutation|Subscription]), so
// a repository can forbid a "get" prefix on Query fields specifically
// without forbidding it everywhere. A directive usage's own arguments
// (Kind.ARGUMENT, for example "reason" in `@deprecated(reason:
// "...")`) are out of scope: no repository's config exercises them
// today.

package lint

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/getoutreach/devbase/v2/internal/graphql/config"
	"github.com/getoutreach/gobox/pkg/set"
	"github.com/vektah/gqlparser/v2/ast"
	"github.com/vektah/gqlparser/v2/gqlerror"
)

// Selector keys naming-convention recognizes beyond the 6 type-kind
// keys in typeDefinitionKindKeys (descriptions.go), matching
// @graphql-eslint's own option keys for the AST kinds that appear in a
// schema file.
const (
	// fieldDefinitionSelector is an object or interface field.
	fieldDefinitionSelector = "FieldDefinition"

	// inputValueDefinitionSelector is a field's or directive
	// definition's own argument, or an input object field.
	inputValueDefinitionSelector = "InputValueDefinition"

	// directiveDefinitionSelector is a directive definition.
	directiveDefinitionSelector = "DirectiveDefinition"

	// enumValueDefinitionSelector is an enum value.
	enumValueDefinitionSelector = "EnumValueDefinition"
)

// rootFieldSelector returns the "FieldDefinition[parent.name.value=...]"
// selector key for a field declared directly on role ("Query",
// "Mutation", or "Subscription"), matching @graphql-eslint's own
// documented "recommended schema" config.
func rootFieldSelector(role string) string {
	return fmt.Sprintf("%s[parent.name.value=%s]", fieldDefinitionSelector, role)
}

// namingStyleRegexes classifies a name (with any leading/trailing
// underscores already stripped) into one of @graphql-eslint's 4
// allowed casing styles, using the exact patterns from its own
// StyleToRegex.
//
//nolint:gochecknoglobals // Why: a fixed lookup table, never mutated; Go has no map consts.
var namingStyleRegexes = map[string]*regexp.Regexp{
	"camelCase":  regexp.MustCompile(`^[a-z][\dA-Za-z]*$`),
	"PascalCase": regexp.MustCompile(`^[A-Z][\dA-Za-z]*$`),
	"snake_case": regexp.MustCompile(`^[a-z][\d_a-z]*[\da-z]*$`),
	"UPPER_CASE": regexp.MustCompile(`^[A-Z][\dA-Z_]*[\dA-Z]*$`),
}

// namingTypeKindDisplay names a type-definition kind's own display
// label in a naming-convention message (for example `Type "b" should
// ...`), matching @graphql-eslint's own KindToDisplayName. This is
// deliberately not typeKindLabel (descriptions.go): that one is
// lowercase for a different message shape (`field "foo" in type
// "Bar"`), not naming-convention's own capitalized, parent-less one.
//
//nolint:gochecknoglobals // Why: a fixed lookup table, never mutated; Go has no map consts.
var namingTypeKindDisplay = map[ast.DefinitionKind]string{
	ast.Scalar:      "Scalar",
	ast.Object:      "Type",
	ast.Interface:   "Interface",
	ast.Union:       "Union",
	ast.Enum:        "Enumerator",
	ast.InputObject: "Input type",
}

// namingRule is one selector's casing/prefix/suffix requirement,
// decoded from either the short form (a bare style string) or the
// long form (an options object) of a naming-convention selector
// value, matching @graphql-eslint's own PropertySchema.
type namingRule struct {
	style             string
	prefix            string
	suffix            string
	forbiddenPrefixes []string
	forbiddenSuffixes []string

	// ignoreRegex is ignorePattern, already compiled once here rather
	// than by every namingRuleMessage call the rule applies to, or nil
	// if ignorePattern was unset or failed to compile.
	ignoreRegex *regexp.Regexp
}

// parseNamingRule decodes v -- a selector's value in
// scripts/devbase.yaml -- into a namingRule. v is either a bare style
// string ("PascalCase") or an options object; any other shape is
// reported as not ok, so the caller can leave the selector
// unconfigured rather than panic on a malformed config.
func parseNamingRule(v any) (namingRule, bool) {
	if style, ok := v.(string); ok {
		return namingRule{style: style}, true
	}
	opts, ok := v.(map[string]any)
	if !ok {
		return namingRule{}, false
	}
	rule := namingRule{
		style:             stringOption(opts, "style"),
		prefix:            stringOption(opts, "prefix"),
		suffix:            stringOption(opts, "suffix"),
		forbiddenPrefixes: stringSliceOption(opts, "forbiddenPrefixes"),
		forbiddenSuffixes: stringSliceOption(opts, "forbiddenSuffixes"),
	}
	if pattern := stringOption(opts, "ignorePattern"); pattern != "" {
		if re, err := regexp.Compile(pattern); err == nil {
			rule.ignoreRegex = re
		}
	}
	return rule, true
}

// stringOption reads a string option named key out of opts, defaulting
// to "" if opts is nil, key is absent, or its value isn't a string.
func stringOption(opts map[string]any, key string) string {
	s, _ := opts[key].(string)
	return s
}

// stringSliceOption reads a string-list option named key out of opts,
// skipping any element that isn't itself a string. It returns nil if
// opts is nil, key is absent, or its value isn't a list.
func stringSliceOption(opts map[string]any, key string) []string {
	raw, ok := opts[key].([]any)
	if !ok {
		return nil
	}
	out := make([]string, 0, len(raw))
	for _, v := range raw {
		if s, ok := v.(string); ok {
			out = append(out, s)
		}
	}
	return out
}

// namingConventionOptions is naming-convention's scripts/devbase.yaml
// options, decoded from RuleConfig.Options.
type namingConventionOptions struct {
	// allowLeadingUnderscore and allowTrailingUnderscore, if false
	// (the default, matching @graphql-eslint's own default), make
	// every name's leading/trailing underscores their own violation,
	// independent of and in addition to whatever selectors below
	// apply to that name.
	allowLeadingUnderscore  bool
	allowTrailingUnderscore bool

	// types is the fallback namingRule for a type-definition site
	// whose own kind key (for example "ObjectTypeDefinition") has no
	// selector entry of its own, or nil if the "types" option was not
	// set.
	types *namingRule

	// selectors holds every other selector key present in
	// scripts/devbase.yaml as-is (for example "FieldDefinition" or
	// "FieldDefinition[parent.name.value=Query]"), decoded to a
	// namingRule. A key this package does not recognize (an
	// operation-only kind, or an arbitrary selector no repository's
	// config uses today -- see this file's package doc comment) is
	// carried here too but never matched against any namingSite, so
	// it is simply inert rather than rejected.
	selectors map[string]namingRule
}

// parseNamingConventionOptions decodes opts (RuleConfig.Options) into
// namingConventionOptions.
func parseNamingConventionOptions(opts map[string]any) namingConventionOptions {
	o := namingConventionOptions{
		allowLeadingUnderscore:  boolOption(opts, "allowLeadingUnderscore"),
		allowTrailingUnderscore: boolOption(opts, "allowTrailingUnderscore"),
	}
	for key, v := range opts {
		if key == "allowLeadingUnderscore" || key == "allowTrailingUnderscore" {
			continue
		}
		rule, ok := parseNamingRule(v)
		if !ok {
			continue
		}
		if key == "types" {
			o.types = &rule
			continue
		}
		if o.selectors == nil {
			o.selectors = make(map[string]namingRule)
		}
		o.selectors[key] = rule
	}
	return o
}

// ruleFor resolves the namingRule that applies to s, if any: s's own
// root-field selector (if it has one) or base selector, whichever
// scripts/devbase.yaml actually configured, falling back to the
// shared "types" rule only for a type-definition site whose own kind
// key was not configured individually.
func (o namingConventionOptions) ruleFor(s *namingSite) (namingRule, bool) {
	if s.rootSelector != "" {
		if r, ok := o.selectors[s.rootSelector]; ok {
			return r, true
		}
	}
	if r, ok := o.selectors[s.baseSelector]; ok {
		return r, true
	}
	if s.typeDefKindKey != "" && o.types != nil {
		return *o.types, true
	}
	return namingRule{}, false
}

// namingSite is one name in a schema file that naming-convention
// checks: a type definition, field, input value (an argument or an
// input object field), enum value, or directive definition.
type namingSite struct {
	pos            *ast.Position
	name           string
	kindLabel      string
	baseSelector   string
	rootSelector   string
	typeDefKindKey string
}

// rootTypeRoles maps each of schema's root operation type names (Query,
// Mutation, and/or Subscription -- whatever schema actually declares)
// to its role, for rootFieldSelector.
func rootTypeRoles(schema *ast.Schema) map[string]string {
	roles := make(map[string]string, 3)
	if schema.Query != nil {
		roles[schema.Query.Name] = "Query"
	}
	if schema.Mutation != nil {
		roles[schema.Mutation.Name] = "Mutation"
	}
	if schema.Subscription != nil {
		roles[schema.Subscription.Name] = "Subscription"
	}
	return roles
}

// namingSites collects every namingSite in doc that was written in one
// of inScope's sources -- excluding gqlparser's built-in prelude and
// any federation- or scalars-synthesized prelude (see federation.go) --
// in document order. Like typenamePrefixViolations, an extension's
// fields are only walked on their own when its base type has no
// definition anywhere; otherwise validator.ValidateSchemaDocument has
// already merged them into the base Definition.Fields that the
// doc.Definitions walk already covers (see descriptions.go's
// forEachDefinition).
func namingSites(doc *ast.SchemaDocument, roles map[string]string, inScope set.Set[*ast.Source]) []namingSite {
	var sites []namingSite

	addIfInScope := func(site namingSite) {
		if site.pos == nil || !inScope.Contains(site.pos.Src) {
			return
		}
		sites = append(sites, site)
	}

	addFields := func(def *ast.Definition) {
		rootSelector := ""
		if role := roles[def.Name]; role != "" {
			rootSelector = rootFieldSelector(role)
		}

		for _, f := range def.Fields {
			if f.Position == nil {
				// __schema/__type introspection meta-fields; see
				// descriptions.go.
				continue
			}
			if def.Kind == ast.InputObject {
				addIfInScope(namingSite{pos: f.Position, name: f.Name, kindLabel: "Input property", baseSelector: inputValueDefinitionSelector})
			} else {
				addIfInScope(namingSite{
					pos: f.Position, name: f.Name, kindLabel: "Field",
					baseSelector: fieldDefinitionSelector, rootSelector: rootSelector,
				})
			}
			for _, arg := range f.Arguments {
				addIfInScope(namingSite{pos: arg.Position, name: arg.Name, kindLabel: "Input property", baseSelector: inputValueDefinitionSelector})
			}
		}
		for _, ev := range def.EnumValues {
			addIfInScope(namingSite{pos: ev.Position, name: ev.Name, kindLabel: "Enumeration value", baseSelector: enumValueDefinitionSelector})
		}
	}

	forEachDefinition(doc, func(def *ast.Definition) {
		if kindKey, ok := typeDefinitionKindKeys[def.Kind]; ok {
			addIfInScope(namingSite{
				pos: def.Position, name: def.Name, kindLabel: namingTypeKindDisplay[def.Kind],
				baseSelector: kindKey, typeDefKindKey: kindKey,
			})
		}
		addFields(def)
	}, addFields)

	for _, dd := range doc.Directives {
		addIfInScope(namingSite{pos: dd.Position, name: dd.Name, kindLabel: "Directive", baseSelector: directiveDefinitionSelector})
		for _, arg := range dd.Arguments {
			addIfInScope(namingSite{pos: arg.Position, name: arg.Name, kindLabel: "Input property", baseSelector: inputValueDefinitionSelector})
		}
	}

	return sites
}

// namingRuleMessage reports the violation message naming-convention
// should raise for name under rule, or "" if name satisfies rule. It
// checks, in order, exactly like @graphql-eslint's own getError: a
// required prefix, a required suffix, every forbidden prefix, every
// forbidden suffix, then the casing style -- stopping at the first
// that applies, against name with any leading/trailing underscores
// stripped (regardless of whether the underscore checks themselves are
// enabled). ignorePattern, if it matches, skips all of the above.
func namingRuleMessage(name string, rule *namingRule) string {
	trimmed := strings.Trim(name, "_")

	if rule.ignoreRegex != nil && rule.ignoreRegex.MatchString(trimmed) {
		return ""
	}
	if rule.prefix != "" && !strings.HasPrefix(trimmed, rule.prefix) {
		return fmt.Sprintf("should have %q prefix", rule.prefix)
	}
	if rule.suffix != "" && !strings.HasSuffix(trimmed, rule.suffix) {
		return fmt.Sprintf("should have %q suffix", rule.suffix)
	}
	for _, p := range rule.forbiddenPrefixes {
		if strings.HasPrefix(trimmed, p) {
			return fmt.Sprintf("should not have %q prefix", p)
		}
	}
	for _, s := range rule.forbiddenSuffixes {
		if strings.HasSuffix(trimmed, s) {
			return fmt.Sprintf("should not have %q suffix", s)
		}
	}
	if rule.style == "" {
		return ""
	}
	if re, ok := namingStyleRegexes[rule.style]; ok && !re.MatchString(trimmed) {
		return fmt.Sprintf("should be in %s format", rule.style)
	}
	return ""
}

// namingConventionViolations reports a RuleNamingConvention violation
// for every site in sites that either has a configured namingRule it
// fails (per namingRuleMessage) or an underscore opts forbids, in
// document order. Both checks run for the same site independently --
// neither one skips the other -- matching @graphql-eslint's own
// behavior of registering them as separate listeners.
func namingConventionViolations(sites []namingSite, opts namingConventionOptions) []Violation {
	var violations []Violation
	for _, s := range sites {
		if rule, ok := opts.ruleFor(&s); ok {
			if msg := namingRuleMessage(s.name, &rule); msg != "" {
				violations = append(violations, Violation{
					err:  gqlerror.ErrorPosf(s.pos, "%s %q %s", s.kindLabel, s.name, msg),
					Rule: config.RuleNamingConvention,
				})
			}
		}
		if !opts.allowLeadingUnderscore && strings.HasPrefix(s.name, "_") {
			violations = append(violations, Violation{
				err:  gqlerror.ErrorPosf(s.pos, "Leading underscores are not allowed"),
				Rule: config.RuleNamingConvention,
			})
		}
		if !opts.allowTrailingUnderscore && strings.HasSuffix(s.name, "_") {
			violations = append(violations, Violation{
				err:  gqlerror.ErrorPosf(s.pos, "Trailing underscores are not allowed"),
				Rule: config.RuleNamingConvention,
			})
		}
	}
	return violations
}

// tier3NamingConvention runs naming-convention against parsed. It does
// not run unless cfg enables it (config.Lint.Enabled).
func tier3NamingConvention(parsed *parsedSchema, inScope set.Set[*ast.Source], cfg *config.Lint) []Violation {
	if !cfg.Enabled(config.RuleNamingConvention) {
		return nil
	}
	sites := namingSites(parsed.doc, rootTypeRoles(parsed.schema), inScope)
	opts := parseNamingConventionOptions(cfg.Options(config.RuleNamingConvention))
	return namingConventionViolations(sites, opts)
}

File: graphql-lint/crates/gql_lint_rules/src/naming_convention.rs
//! Ports `internal/graphql/lint/naming_convention.go`'s `naming-convention`
//! rule: requires names to follow the casing, prefix/suffix, and
//! underscore conventions its options configure per kind of definition.
//!
//! Like the Go port, this narrows `@graphql-eslint`'s own option surface
//! (which accepts any AST node kind, including operation-only kinds that
//! never appear in a standalone schema file) to the kinds that can occur
//! in one: the 6 type-definition kinds (via a shared `types` fallback and
//! per-kind overrides), `FieldDefinition`, `InputValueDefinition` (an
//! argument or an input object field), `EnumValueDefinition`, and
//! `DirectiveDefinition` — plus the 3 root-operation-type field selectors
//! from `@graphql-eslint`'s own documented "recommended schema" config
//! (`FieldDefinition[parent.name.value=Query|Mutation|Subscription]`), so a
//! repo can forbid a `get` prefix on `Query` fields specifically.

use apollo_compiler::Schema;
use apollo_compiler::schema::{ExtendedType, Name};
use gql_lint_core::{Violation, rules};
use regex::Regex;
use std::collections::HashMap;
use std::sync::LazyLock;

/// The 4 casing styles `@graphql-eslint` recognizes, with the exact
/// patterns its `StyleToRegex` uses.
static NAMING_STYLE_REGEXES: LazyLock<HashMap<&'static str, Regex>> = LazyLock::new(|| {
    [
        ("camelCase", r"^[a-z][0-9A-Za-z]*$"),
        ("PascalCase", r"^[A-Z][0-9A-Za-z]*$"),
        ("snake_case", r"^[a-z][0-9_a-z]*[0-9a-z]*$"),
        ("UPPER_CASE", r"^[A-Z][0-9A-Z_]*[0-9A-Z]*$"),
    ]
    .into_iter()
    .map(|(name, pattern)| (name, Regex::new(pattern).expect("built-in pattern")))
    .collect()
});

const SCALAR_TYPE_DEFINITION: &str = "ScalarTypeDefinition";
const OBJECT_TYPE_DEFINITION: &str = "ObjectTypeDefinition";
const INTERFACE_TYPE_DEFINITION: &str = "InterfaceTypeDefinition";
const UNION_TYPE_DEFINITION: &str = "UnionTypeDefinition";
const ENUM_TYPE_DEFINITION: &str = "EnumTypeDefinition";
const INPUT_OBJECT_TYPE_DEFINITION: &str = "InputObjectTypeDefinition";
const FIELD_DEFINITION: &str = "FieldDefinition";
const INPUT_VALUE_DEFINITION: &str = "InputValueDefinition";
const DIRECTIVE_DEFINITION: &str = "DirectiveDefinition";
const ENUM_VALUE_DEFINITION: &str = "EnumValueDefinition";

fn root_field_selector(role: &str) -> String {
    format!("{FIELD_DEFINITION}[parent.name.value={role}]")
}

/// One selector's casing/prefix/suffix requirement, decoded from either
/// the short form (a bare style string) or the long form (an options
/// mapping) of a `naming-convention` selector value.
#[derive(Debug, Clone, Default)]
pub struct NamingRule {
    pub style: Option<String>,
    pub prefix: Option<String>,
    pub suffix: Option<String>,
    pub forbidden_prefixes: Vec<String>,
    pub forbidden_suffixes: Vec<String>,
    pub ignore_regex: Option<Regex>,
}

impl NamingRule {
    fn from_yaml(value: &serde_yaml::Value) -> Option<Self> {
        if let Some(style) = value.as_str() {
            return Some(Self {
                style: Some(style.to_string()),
                ..Self::default()
            });
        }
        let mapping = value.as_mapping()?;
        let string_opt = |key: &str| {
            mapping
                .get(key)
                .and_then(serde_yaml::Value::as_str)
                .map(String::from)
        };
        let string_list_opt = |key: &str| -> Vec<String> {
            mapping
                .get(key)
                .and_then(serde_yaml::Value::as_sequence)
                .map(|seq| {
                    seq.iter()
                        .filter_map(serde_yaml::Value::as_str)
                        .map(String::from)
                        .collect()
                })
                .unwrap_or_default()
        };
        Some(Self {
            style: string_opt("style"),
            prefix: string_opt("prefix"),
            suffix: string_opt("suffix"),
            forbidden_prefixes: string_list_opt("forbiddenPrefixes"),
            forbidden_suffixes: string_list_opt("forbiddenSuffixes"),
            ignore_regex: string_opt("ignorePattern").and_then(|p| Regex::new(&p).ok()),
        })
    }
}

/// `naming-convention`'s `scripts/devbase.yaml` options, decoded from the
/// rule's raw options mapping.
#[derive(Debug, Clone, Default)]
pub struct NamingConventionOptions {
    /// If false (the default, matching `@graphql-eslint`'s own default),
    /// a leading underscore is its own violation, independent of whatever
    /// selector rule also applies to that name.
    pub allow_leading_underscore: bool,
    pub allow_trailing_underscore: bool,
    /// Fallback rule for a type-definition site whose own kind key has no
    /// selector entry of its own.
    pub types: Option<NamingRule>,
    /// Every other selector key present in the config, as-is (for example
    /// `"FieldDefinition"` or `"FieldDefinition[parent.name.value=Query]"`).
    pub selectors: HashMap<String, NamingRule>,
}

impl NamingConventionOptions {
    #[must_use]
    pub fn from_yaml(options: Option<&serde_yaml::Mapping>) -> Self {
        let Some(options) = options else {
            return Self::default();
        };

        let mut opts = Self {
            allow_leading_underscore: options
                .get("allowLeadingUnderscore")
                .and_then(serde_yaml::Value::as_bool)
                .unwrap_or(false),
            allow_trailing_underscore: options
                .get("allowTrailingUnderscore")
                .and_then(serde_yaml::Value::as_bool)
                .unwrap_or(false),
            types: None,
            selectors: HashMap::new(),
        };

        for (key, value) in options {
            let Some(key) = key.as_str() else { continue };
            if key == "allowLeadingUnderscore" || key == "allowTrailingUnderscore" {
                continue;
            }
            let Some(rule) = NamingRule::from_yaml(value) else {
                continue;
            };
            if key == "types" {
                opts.types = Some(rule);
            } else {
                opts.selectors.insert(key.to_string(), rule);
            }
        }

        opts
    }

    /// Resolves the rule that applies to `site`: its own root-field
    /// selector (if any) or base selector, whichever the config actually
    /// set, falling back to the shared `types` rule only for a
    /// type-definition site whose own kind key was not configured
    /// individually.
    fn rule_for<'a>(&'a self, site: &NamingSite) -> Option<&'a NamingRule> {
        if let Some(root_selector) = &site.root_selector
            && let Some(rule) = self.selectors.get(root_selector)
        {
            return Some(rule);
        }
        if let Some(rule) = self.selectors.get(site.base_selector) {
            return Some(rule);
        }
        if site.type_def_kind_key.is_some() {
            return self.types.as_ref();
        }
        None
    }
}

/// One name in a schema that `naming-convention` checks.
struct NamingSite<'a> {
    name: &'a Name,
    kind_label: &'static str,
    base_selector: &'static str,
    root_selector: Option<String>,
    type_def_kind_key: Option<&'static str>,
}

/// Runs the `naming-convention` rule over `schema`, per `opts`.
#[must_use]
pub fn naming_convention(schema: &Schema, opts: &NamingConventionOptions) -> Vec<Violation> {
    let roles = root_type_roles(schema);
    let sites = collect_sites(schema, &roles);

    let mut violations = Vec::with_capacity(sites.len());
    for site in &sites {
        if let Some(rule) = opts.rule_for(site)
            && let Some(msg) = naming_rule_message(site.name.as_str(), rule)
        {
            violations.push(violation(schema, site, &msg));
        }
        if !opts.allow_leading_underscore && site.name.as_str().starts_with('_') {
            violations.push(violation(
                schema,
                site,
                "Leading underscores are not allowed",
            ));
        }
        if !opts.allow_trailing_underscore && site.name.as_str().ends_with('_') {
            violations.push(violation(
                schema,
                site,
                "Trailing underscores are not allowed",
            ));
        }
    }
    violations
}

fn violation(schema: &Schema, site: &NamingSite, message: &str) -> Violation {
    let (file, line, column) =
        gql_lint_core::resolve_location(&schema.sources, site.name.location());
    Violation {
        file,
        line,
        column,
        message: format!("{} \"{}\" {}", site.kind_label, site.name, message),
        rule: rules::NAMING_CONVENTION,
    }
}

/// Maps each of `schema`'s root operation type names to its role
/// ("Query", "Mutation", "Subscription"), for [`root_field_selector`].
fn root_type_roles(schema: &Schema) -> HashMap<&str, &'static str> {
    let mut roles = HashMap::with_capacity(3);
    let def = &schema.schema_definition;
    if let Some(name) = &def.query {
        roles.insert(name.as_str(), "Query");
    }
    if let Some(name) = &def.mutation {
        roles.insert(name.as_str(), "Mutation");
    }
    if let Some(name) = &def.subscription {
        roles.insert(name.as_str(), "Subscription");
    }
    roles
}

/// Collects every [`NamingSite`] in `schema`, excluding built-in types —
/// see [`crate::alphabetize::alphabetize`]'s doc comment for why.
fn collect_sites<'a>(
    schema: &'a Schema,
    roles: &HashMap<&str, &'static str>,
) -> Vec<NamingSite<'a>> {
    let mut sites = Vec::new();

    for ty in gql_lint_core::in_scope_types(schema) {
        let (kind_label, kind_key, name) = match ty {
            ExtendedType::Scalar(t) => ("Scalar", SCALAR_TYPE_DEFINITION, &t.name),
            ExtendedType::Object(t) => ("Type", OBJECT_TYPE_DEFINITION, &t.name),
            ExtendedType::Interface(t) => ("Interface", INTERFACE_TYPE_DEFINITION, &t.name),
            ExtendedType::Union(t) => ("Union", UNION_TYPE_DEFINITION, &t.name),
            ExtendedType::Enum(t) => ("Enumerator", ENUM_TYPE_DEFINITION, &t.name),
            ExtendedType::InputObject(t) => ("Input type", INPUT_OBJECT_TYPE_DEFINITION, &t.name),
        };
        sites.push(NamingSite {
            name,
            kind_label,
            base_selector: kind_key,
            root_selector: None,
            type_def_kind_key: Some(kind_key),
        });

        match ty {
            ExtendedType::Object(t) => {
                add_fields(&mut sites, t.name.as_str(), t.fields.values(), roles);
            }
            ExtendedType::Interface(t) => {
                add_fields(&mut sites, t.name.as_str(), t.fields.values(), roles);
            }
            ExtendedType::InputObject(t) => {
                for field in t.fields.values() {
                    sites.push(NamingSite {
                        name: &field.name,
                        kind_label: "Input property",
                        base_selector: INPUT_VALUE_DEFINITION,
                        root_selector: None,
                        type_def_kind_key: None,
                    });
                }
            }
            ExtendedType::Enum(t) => {
                for value in t.values.values() {
                    sites.push(NamingSite {
                        name: &value.value,
                        kind_label: "Enumeration value",
                        base_selector: ENUM_VALUE_DEFINITION,
                        root_selector: None,
                        type_def_kind_key: None,
                    });
                }
            }
            _ => {}
        }
    }

    for directive in gql_lint_core::in_scope_directive_definitions(schema) {
        sites.push(NamingSite {
            name: &directive.name,
            kind_label: "Directive",
            base_selector: DIRECTIVE_DEFINITION,
            root_selector: None,
            type_def_kind_key: None,
        });
        for arg in &directive.arguments {
            sites.push(NamingSite {
                name: &arg.name,
                kind_label: "Input property",
                base_selector: INPUT_VALUE_DEFINITION,
                root_selector: None,
                type_def_kind_key: None,
            });
        }
    }

    sites
}

fn add_fields<'a>(
    sites: &mut Vec<NamingSite<'a>>,
    type_name: &str,
    fields: impl Iterator<
        Item = &'a apollo_compiler::schema::Component<apollo_compiler::schema::FieldDefinition>,
    >,
    roles: &HashMap<&str, &'static str>,
) {
    let root_selector = roles.get(type_name).map(|role| root_field_selector(role));
    for field in fields {
        sites.push(NamingSite {
            name: &field.name,
            kind_label: "Field",
            base_selector: FIELD_DEFINITION,
            root_selector: root_selector.clone(),
            type_def_kind_key: None,
        });
        for arg in &field.arguments {
            sites.push(NamingSite {
                name: &arg.name,
                kind_label: "Input property",
                base_selector: INPUT_VALUE_DEFINITION,
                root_selector: None,
                type_def_kind_key: None,
            });
        }
    }
}

/// Reports the violation message `naming-convention` should raise for
/// `name` under `rule`, or `None` if `name` satisfies it. Checks, in
/// order (matching `@graphql-eslint`'s own `getError`): a required
/// prefix, a required suffix, every forbidden prefix, every forbidden
/// suffix, then the casing style — stopping at the first one that
/// applies, against `name` with leading/trailing underscores stripped.
/// `ignorePattern`, if it matches, skips all of the above.
fn naming_rule_message(name: &str, rule: &NamingRule) -> Option<String> {
    let trimmed = name.trim_matches('_');

    if let Some(ignore) = &rule.ignore_regex
        && ignore.is_match(trimmed)
    {
        return None;
    }
    if let Some(prefix) = &rule.prefix
        && !trimmed.starts_with(prefix.as_str())
    {
        return Some(format!("should have \"{prefix}\" prefix"));
    }
    if let Some(suffix) = &rule.suffix
        && !trimmed.ends_with(suffix.as_str())
    {
        return Some(format!("should have \"{suffix}\" suffix"));
    }
    for forbidden in &rule.forbidden_prefixes {
        if trimmed.starts_with(forbidden.as_str()) {
            return Some(format!("should not have \"{forbidden}\" prefix"));
        }
    }
    for forbidden in &rule.forbidden_suffixes {
        if trimmed.ends_with(forbidden.as_str()) {
            return Some(format!("should not have \"{forbidden}\" suffix"));
        }
    }
    let style = rule.style.as_deref()?;
    let re = NAMING_STYLE_REGEXES.get(style)?;
    if re.is_match(trimmed) {
        None
    } else {
        Some(format!("should be in {style} format"))
    }
}
